packages/langium/src/node/node-file-system-provider.ts

```diff
@@ -0,0 +1,51 @@
+/******************************************************************************
+ * Copyright 2022 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+
+import fs from 'fs';
+import { URI, Utils } from 'vscode-uri';
+import { FileSystemFilter, FileSystemNode, FileSystemProvider } from '../workspace/file-system-provider';
+
+export class NodeFileSystemProvider implements FileSystemProvider {
+
+    readFile(uri: URI): Promise<string> {
+        return fs.promises.readFile(uri.fsPath, 'utf-8');
+    }
+
+    readFileSync(uri: URI): string {
+        return fs.readFileSync(uri.fsPath, 'utf-8');
+    }
+
+    async traverse(root: URI, filter?: FileSystemFilter): Promise<URI[]> {
```

Comment:
Proposal: expose a callback function instead of returning an array in the API. That way the method is more flexible, for example you don't need to `flat()` the workspace folder entries in the DefaultWorkspaceManager.